### PR TITLE
Open engine should keep only starting commit

### PR DESCRIPTION
**Pull request by dnhatn**
Keeping existing unsafe commits when opening an engine can be problematic because these commits are not safe at the recovering time but they can suddenly become safe in the future.
The following issues can happen if unsafe commits are kept oninit.

**1. Replica can use unsafe commit in peer-recovery**. This happens when a replica with a safe commit c1(max_seqno=1) and an unsafe commit c2(max_seqno=2) recovers from a primary with c1(max_seqno=1). If a new document(seqno=2) is added without flushing, the global checkpoint is advanced to 2; and the replica recovers again, it will use the unsafe commit c2(max_seqno=2 <= gcp=2) as the starting commit for sequenced-based recovery even the commit c2 contains a stale operation and the document(with seqno=2) will not be replicated to the replica.

**2. Min translog gen for recovery can go backwards in peer-recovery.** This happens when are replica with a safe commit c1(local_checkpoint=1, recovery_translog_gen=1) and an unsafe commit c2(local_checkpoint=2, recovery_translog_gen=2). The replica recovers from a primary, and keeps c2 as the last commit, then sets last_translog_gen to 2. Flushing a new commit on the replica will cause exception as the new last commit c3 will have recovery_translog_gen=1. The recovery translog generation of a commit is calculated based on the current local checkpoint. The local checkpoint of c3 is 1 while the local checkpoint of c2 is 2.

**3. Commit without translog can be used for recovery**. An old index, which was created before multiple-commits is introduced (v6.2), may not have a safe commit. If that index has a snapshotted commit without translog and an unsafe commit, the policy can consider the snapshotted commit as a safe commit for recovery even the commit does not have translog.

These issues can be avoided if the combined deletion policy keeps only the starting commit onInit.

Relates #27804
Relates #28181

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: since we're changing the logic of `OPEN_INDEX_AND_TRANSLOG`, I think we can also consolidate the code paths for `OPEN_INDEX_CREATE_TRANSLOG` & `OPEN_INDEX_AND_TRANSLOG` to be the same whenever we open an index - less exceptions and more unity. To do so we need to set `startingCommit` to a valid value in `InternalEngine#getStartingCommitPoint` whenever we are actually opening something.

File: server/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java]: can you add java docs explaining why this behavior is needed. I think it will be hard to figure out.

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: can you leave a big comment explaining why this is needed?

**Review by bleskes — CHANGES_REQUESTED** [review]
I think this makes sense. I left a suggestion. Also, I think it will be good to have an engine level test to show that opening an index cleans up commits above the safe commit. Last, we discussed some clarification for the PR description. It will help future readers to incorporate those in the PR description.

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: This feels strange here. Move it to the constructor?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we assert there is only one? I believe this is the case.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I know we now trim things and this OK, but it feels risky - maybe we'll change our mind later in the future clean less heavily. It will be hard to find this place and change. Maybe just keep it as is?

**Review by bleskes — APPROVED** [review]
LGTM. I left some minor feedback. No need for another review on my side.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: The starting commit will be deleted before this line is executed if the open mode is `OPEN_INDEX_CREATE_TRANSLOG` as we have created a new commit (refresh translog uuid + gen).

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I added an assert but moved to `IndexShard#openIndexAndCreateTranslog`. We have some existing tests which opens engine with multiple commits.

**Comment by dnhatn** [comment]
@bleskes Thank you for your helpful reviews. I will wait for CI then merge.

**Pull request merged**
